File: web/src/lib/listKind.test.ts
import { describe, expect, it } from "vitest";
import { isExpenses, listKind, listKindIcon, listKindLabelKey, showsShoppingFields } from "./listKind";
import type { ListObject } from "../api/contract";

function list(kind?: string): ListObject {
  const clock = { updated_at: 1, updated_by: "dev" };
  return {
    id: "list-1",
    created_at: 0,
    fields: {
      name: { value: "L", ...clock },
      ...(kind ? { kind: { value: kind as "shopping" | "checklist", ...clock } } : {}),
    },
  };
}

describe("listKind (T-110)", () => {
  it("defaults to shopping when the field is absent", () => {
    // Pre-T-110 rows and older servers send no kind at all — they must keep behaving as before.
    expect(listKind(list())).toBe("shopping");
    expect(listKind(undefined)).toBe("shopping");
  });

  it("reads an explicit kind", () => {
    expect(listKind(list("checklist"))).toBe("checklist");
    expect(listKind(list("shopping"))).toBe("shopping");
  });

  it("only shopping lists show the shopping-only fields", () => {
    expect(showsShoppingFields("shopping")).toBe(true);
    expect(showsShoppingFields("checklist")).toBe(false);
    expect(showsShoppingFields("expenses")).toBe(false);
  });

  it("recognises an expenses list", () => {
    expect(isExpenses("expenses")).toBe(true);
    expect(isExpenses("shopping")).toBe(false);
    expect(isExpenses("checklist")).toBe(false);
  });

  it("labels and icons the three kinds distinctly", () => {
    const keys = (["shopping", "checklist", "expenses"] as const).map(listKindLabelKey);
    const icons = (["shopping", "checklist", "expenses"] as const).map(listKindIcon);
    expect(new Set(keys).size).toBe(3);
    expect(new Set(icons).size).toBe(3);
  });
});

File: android/app/src/test/java/org/p23q/shoppinglist/data/ListKindTest.kt
package org.p23q.shoppinglist.data

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * T-110 list-kind rules. Must stay in lockstep with the web client's listKind.test.ts — the two
 * clients share the rule (client-ui-notes.md) so they render the same synced list identically.
 */
class ListKindTest {

    @Test
    fun `absent or unknown kinds fall back to shopping`() {
        // Pre-T-110 rows and older servers send no kind; a newer client could send one we don't know.
        assertEquals(ListKind.SHOPPING, ListKind.of(null))
        assertEquals(ListKind.SHOPPING, ListKind.of(""))
        assertEquals(ListKind.SHOPPING, ListKind.of("tasks"))
        assertEquals(ListKind.SHOPPING, ListKind.DEFAULT)
    }

    @Test
    fun `explicit kinds round-trip`() {
        assertEquals(ListKind.CHECKLIST, ListKind.of("checklist"))
        assertEquals(ListKind.SHOPPING, ListKind.of("shopping"))
    }

    @Test
    fun `only shopping lists show the shopping-only fields`() {
        assertTrue(ListKind.showsShoppingFields(ListKind.SHOPPING))
        assertTrue(ListKind.showsShoppingFields(null)) // default stays a shopping list
        assertFalse(ListKind.showsShoppingFields(ListKind.CHECKLIST))
    }

    @Test
    fun `labels and icons distinguish the two kinds`() {
        assertEquals("Shopping list", ListKind.label(ListKind.SHOPPING))
        assertEquals("Checklist", ListKind.label(ListKind.CHECKLIST))
        assertTrue(ListKind.icon(ListKind.SHOPPING) != ListKind.icon(ListKind.CHECKLIST))
    }
}
